src/tools/fuzzing/heap-types.cpp

```diff
@@ -80,35 +86,68 @@ struct HeapTypeGeneratorImpl {
       }
     }
 
+    // Initialize the recursion groups.
+    recGroupEnds.reserve(builder.size());
+    if (getTypeSystem() != TypeSystem::Isorecursive) {
+      // Recursion groups don't matter and we can choose children as though we
+      // had a single large recursion group.
+      for (Index i = 0; i < builder.size(); ++i) {
+        recGroupEnds.push_back(builder.size());
+      }
+    } else {
+      // We are using isorecursive types, so create groups. Choose an expected
+      // group size uniformly at random, then create groups with random sizes on
+      // a geometric distribution based on that expected size.
+      size_t expectedSize = 1 + rand.upTo(builder.size());
+      Index groupStart = 0;
+      for (Index i = 0; i < builder.size(); ++i) {
+        if (rand.oneIn(expectedSize)) {
+          // End the old group and create a new group.
+          Index newGroupStart = i + 1;
+          builder.createRecGroup(groupStart, newGroupStart - groupStart);
+          for (Index j = groupStart; j < newGroupStart; ++j) {
+            recGroupEnds.push_back(newGroupStart);
+          }
+          groupStart = newGroupStart;
+        }
+      }
+      // Finish the last group.
+      builder.createRecGroup(groupStart, builder.size() - groupStart);
+      for (Index j = groupStart; j < builder.size(); ++j) {
+        recGroupEnds.push_back(builder.size());
+      }
```

Comment:
I think you can avoid the extra code on the outside for the last group by doing a more efficient loop that picks group sizes, rather than tests on each item for "is end of group", like this:
```python
position = 0
while position < builder.size():
  curr_size = std::min(upTo(groupSize), builder.size())
  add_group([position, position + curr_size])
  position += curr_size
```
But if you have a preference for your current loop, how about replacing `if (rand.oneIn(expectedSize))` with `if (rand.oneIn(expectedSize) || i == builder.size() - 1)`?

Reply:
I like the current loop because it gives a nice geometric distribution for the size of the group, but I'll see if I can deduplicate the code.